exchanges/ftx/ftx_wrapper.go

```diff
@@ -0,0 +1,873 @@
+	var resp account.Holdings
+	data, err := f.GetBalances()
+	if err != nil {
+		return resp, err
+	}
+	var acc account.SubAccount
+	for i := range data {
+		c := currency.NewCode(data[i].Coin)
+		hold := data[i].Total - data[i].Free
+		total := data[i].Total
+		acc.Currencies = append(acc.Currencies,
+			account.Balance{CurrencyName: c,
+				TotalValue: total,
+				Hold:       hold})
+	}
+	resp.Accounts = append(resp.Accounts, acc)
+	resp.Exchange = f.Name
+
+	err = account.Process(&resp)
+	if err != nil {
+		return account.Holdings{}, err
+	}
+
+	return resp, nil
+}
+
+// FetchAccountInfo retrieves balances for all enabled currencies
+func (f *FTX) FetchAccountInfo() (account.Holdings, error) {
+	acc, err := account.GetHoldings(f.Name)
+	if err != nil {
+		return f.UpdateAccountInfo()
+	}
+
+	return acc, nil
+}
+
+// GetFundingHistory returns funding history, deposits and
+// withdrawals
+func (f *FTX) GetFundingHistory() ([]exchange.FundHistory, error) {
+	var resp []exchange.FundHistory
+	depositData, err := f.FetchDepositHistory()
+	if err != nil {
+		return resp, err
+	}
+	for x := range depositData {
+		var tempData exchange.FundHistory
+		tempData.Fee = depositData[x].Fee
+		tempData.Timestamp = depositData[x].Time
+		tempData.ExchangeName = f.Name
+		tempData.CryptoTxID = depositData[x].TxID
+		tempData.Status = depositData[x].Status
+		tempData.Amount = depositData[x].Size
+		tempData.Currency = depositData[x].Coin
+		tempData.TransferID = strconv.FormatInt(depositData[x].ID, 10)
+		resp = append(resp, tempData)
+	}
+	withdrawalData, err := f.FetchWithdrawalHistory()
+	if err != nil {
+		return resp, err
+	}
+	for y := range withdrawalData {
+		var tempData exchange.FundHistory
+		tempData.Fee = depositData[y].Fee
+		tempData.Timestamp = depositData[y].Time
+		tempData.ExchangeName = f.Name
+		tempData.CryptoTxID = depositData[y].TxID
+		tempData.Status = depositData[y].Status
+		tempData.Amount = depositData[y].Size
+		tempData.Currency = depositData[y].Coin
+		tempData.TransferID = strconv.FormatInt(depositData[y].ID, 10)
+		resp = append(resp, tempData)
+	}
+	return resp, nil
+}
+
+// GetExchangeHistory returns historic trade data within the timeframe provided.
+func (f *FTX) GetExchangeHistory(p currency.Pair, assetType asset.Item, timestampStart, timestampEnd time.Time) ([]exchange.TradeHistory, error) {
+	marketName := f.FormatExchangeCurrency(p, assetType).String()
+	var resp []exchange.TradeHistory
+	trades, err := f.GetTrades(marketName, time.Unix(timestampStart.Unix(), 0), time.Unix(timestampEnd.Unix(), 0), 100)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		var tempResp exchange.TradeHistory
+		if len(trades) > 0 {
+			tempResp.Amount = trades[0].Size
+			tempResp.Price = trades[0].Price
+			tempResp.Exchange = f.Name
+			tempResp.Timestamp = trades[0].Time
+			tempResp.TID = strconv.FormatInt(trades[0].ID, 10)
+			tempResp.Side = trades[0].Side
+			resp = append(resp, tempResp)
+		}
+		for y := 1; y < len(trades); y++ {
```

Comment:
There is an issue with things here. I'm getting duplicate entries in the response.
One of the possible reasons is that you are not creating a new `var tempResp exchange.TradeHistory` any time you append it. Inside this loop you should create a new one each time.

Otherwise, you may have a problem with your time loop and may be requesting the same data and not checking if you've already added it.

See this gist: https://gist.github.com/gloriousCode/af883b4a484059ef906316faf5fa9a8b
In the comment, I paste an example which has showed up in the response data 9 times

If you want to verify this yourself, in `exchange_wrapper_issues`, change `r8` to the following: `r8, err = e.GetExchangeHistory(p, assetTypes[i], time.Now().Add(-time.Minute * 10 ), time.Now())` and run `exchange_wrapper_issues --exchanges=ftx` and view the report